clusterloader2/pkg/measurement/manager.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package measurement
+
+import (
+	"sync"
+
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+// MeasurementManager manages all measurement executions.
+type MeasurementManager struct {
+	factory   *MeasurementFactory
+	clientSet clientset.Interface
+
+	lock sync.Mutex
+	// measurements is a map from method type and identifier to measurement instance.
+	measurements map[string]map[string]Measurement
+}
+
+// CreateMeasurementManager creates new instance of MeasurementManager.
+func CreateMeasurementManager(factory *MeasurementFactory, clientSet clientset.Interface) *MeasurementManager {
+	return &MeasurementManager{
+		factory:      factory,
```

Comment:
This way users don't need to worry about managing 2 objects (i.e both factory and manager), but only the latter.

Reply:
Infact, you could just merge factory and manager into single struct, as each of them pretty much holds 1 variable.

Reply:
> Infact, you could just merge factory and manager into single struct, as each of them pretty much holds 1 variable.

TL;DR: I will be very unhappy to merge those two structures.
Those two are two separated concept. Factory should be something on global level, that is aware of how to create measurement instances.
Manager exists in some context (test context in this case) - it uses factory to create measurement instances. Managers also stores states of instances.

> Initialize a new instance of factory in CreateMeasurementManager (so you don't need that global variable anymore)
Make Register() a public method on the MeasurementManager and make the Register() for the factory a private one, and call the latter from the former

Could you explain to me, how will measurement methods will register in factory? I still missing the point where Register is called by the measurement method implementation or measurement is added to manager.

With this implementation I try to do similar thing to how apis register to legacyscheme.

Reply:
If factory is something that is really independent of manager, why are you holding it as a field in the manager? So I think it makes sense to either:

- Keep the factory independent of manager and hold them separately, or
- Make manager manage everything. By this I don't mean merging both structs (that was just another suggestion), but I meant that from the cluster-loader perspective it only talks to manager

> Could you explain to me, how will measurement methods will register in factory?

How would it be done currently? It'll be just the same, but instead of calling Register() on factory, you'll call it on manager which will then be responsible for storing it.

Reply:
> If factory is something that is really independent of manager, why are you holding it as a field in the manager?

Factory is something independent. It's manager that depends on factory. Managers uses factory the same way it uses clientSet.
Plus, I can easily imagine having factory initialize once and then have multiple tests run with separate manager each that use this one factory. Therefore is think those two thing are separate concepts.

Reply:
> How would it be done currently? It'll be just the same, but instead of calling Register() on factory, you'll call it on manager which will then be responsible for storing it.

I'm keeping one instance of factory as global variable so measurements can register themselves. Does it mean that using your approach I should make one global manager? Or after creating manager should I explicitly register each measurement? - This part is still not clear for me :(

Reply:
- If factory is independent of manager's lifecycle, then why are your storing it as a field in the manager?
- If factory can exist without manager, why are you making createMeasurement() private? Because that's also something a user of factory would want besides Register()

> Therefore is think those two thing are separate concepts.

I agree and am not saying they're not. It's about what really do you want to expose to users of this package.

Reply:
> I'm keeping one instance of factory as global variable so measurements can register themselves. Does it mean that using your approach I should make one global manager?

Yes. The only change you should need to make is having a manager instance instead of a manager and a factory instance (whichever way you're defining them.. global or local or field ..)

Reply:
> If factory is independent of manager's lifecycle, then why are your storing it as a field in the manager?

Manager is using it therefore pointer is needed. In theory, someone can have separate factory instance. So there should be a way to choose which factory should be used.

TBH, I still prefer to have factory singleton in measurement module (as it was done in first place). This way there is no need to pass factory pointers etc.

> If factory can exist without manager, why are you making createMeasurement() private? Because that's also something a user of factory would want besides Register()

That's good observation. I should make createMeasurement public.

> Yes. The only change you should need to make is having a manager instance instead of a manager and a factory instance (whichever way you're defining them.. global or local or field ..)

Manager is something that should exist only on test level, thus can't be made global. On the other hand factory in independent of any test state so this one can be made global.